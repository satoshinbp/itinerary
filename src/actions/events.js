import db from '../firebase/firebase'

// ADD_EVENT
export const addEvent = event => ({
  type: 'ADD_EVENT',
  event
})

export const startAddEvent = (tripId, eventData) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    const eventsRef = db.collection('users').doc(uid).collection('trips').doc(tripId).collection('events')
    const { title, date, startTime, endTime, location, note } = eventData
    const event = { title, date, startTime, endTime, location, note, user: uid, tripId }

    return eventsRef.add(event).then(snapshot => (
      dispatch(addEvent({
        id: snapshot.id,
        ...event
      }))
    ))
  }
}

// EDIT_EVENT
export const editEvent = (id, updates) => ({
  type: 'EDIT_EVENT',
  id,
  updates
})

export const startEditEvent = (tripId, id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    const eventsRef = db.collection('users').doc(uid).collection('trips').doc(tripId).collection('events').doc(id)
    return eventsRef.update(updates).then(() => dispatch(editEvent(id, updates)))
  }
}

// REMOVE_EVENT
export const removeEvent = id => ({
  type: 'REMOVE_EVENT',
  id
})

export const startRemoveEvent = (tripId, id) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    const eventsRef = db.collection('users').doc(uid).collection('trips').doc(tripId).collection('events').doc(id)
    return eventsRef.delete().then(() => dispatch(removeEvent(id)))
  }
}

// SET_EVENTS
export const setEvents = events => ({
  type: 'SET_EVENTS',
  events
})

export const startSetEvents = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    const events = []

    return db.collectionGroup('events').where("user", "==", uid).get().then(querySnapshot => {
      querySnapshot.forEach(snapshot => {
        events.push({
          id: snapshot.id,
          tripId: snapshot.ref.parent.parent.id,
          user: uid,
          ...snapshot.data()
        })
      })
      dispatch(setEvents(events))
    })
  }
}