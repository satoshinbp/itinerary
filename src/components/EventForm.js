import React, { useState } from 'react'
import { SingleDatePicker } from 'react-dates'
import 'react-dates/lib/css/_datepicker.css'
import 'react-dates/initialize'
import DateFnsUtils from '@date-io/date-fns'
import { TimePicker, MuiPickersUtilsProvider} from '@material-ui/pickers'
import moment from 'moment'

export default (props) => {
  const id = props.event ? props.event.id : props.id
  const tripId = props.event ? props.event.tripId : props.tripId
  const [title, setTitle] = useState(props.event ? props.event.title : '')
  const [date, setDate] = useState(props.event ? moment(props.event.date) : moment(props.date))
  const [startTime, setStartTime] = useState(props.event ? moment(props.event.startTime) : moment(props.date).hour(12).minutes(0))
  const [endTime, setEndTime] = useState(props.event ? moment(props.event.endTime) : moment(props.date).hour(12).minutes(0))
  const [location, setLocation] = useState(props.event ? props.event.location : '')
  const [note, setNote] = useState(props.event ? props.event.note : '')
  const [focused, setFocused] = useState(null)
  const [error, setError] = useState('')

  const onStartTimeChange = time => {
    setStartTime(moment(time))
    if (moment(time).isAfter(endTime)) {
      setEndTime(moment(time))
    }
  }
  const onEndTimeChange = time => {
    setEndTime(moment(time))
    if (moment(time).isBefore(startTime)) {
      setStartTime(moment(time))
    }
  }
  const onSubmit = e => {
    e.preventDefault()

    if (!title) {
      setError('Please provide title.')
    } else {
      setError('')
      props.onSubmit({
        id,
        tripId,
        title,
        date: date.valueOf(),
        startTime: startTime.valueOf(),
        endTime: endTime.valueOf(),
        location,
        note
      })
    }
  }

  return (
    <form className="form" onSubmit={onSubmit}>
      {error && <p className="form__error">{error}</p>}
      <input
        type="text"
        placeholder="title"
        autoFocus
        className="text-input"
        value={title}
        onChange={e => setTitle(e.target.value)}
      />
      <div className="date-time-picker">
        <SingleDatePicker
          date={date}
          onDateChange={date => setDate(moment(date).startOf('day'))}
          focused={focused}
          onFocusChange={({ focused }) => setFocused(focused)}
          isOutsideRange={() => false}
          displayFormat="YYYY/MM/DD"
        />
      </div>
      <div className="date-time-picker">
        <MuiPickersUtilsProvider utils={DateFnsUtils}>
          <TimePicker value={startTime} onChange={onStartTimeChange} label="Start Time" />
        </MuiPickersUtilsProvider>
        <MuiPickersUtilsProvider utils={DateFnsUtils}>
          <TimePicker value={endTime} onChange={onEndTimeChange} label="End Time" />
        </MuiPickersUtilsProvider>
      </div>
      <input
        type="text"
        placeholder="location"
        className="text-input"
        value={location}
        onChange={e => setLocation(e.target.value)}
      />
      <textarea
        placeholder="Add a note for your event (optional)"
        className="textarea"
        value={note}
        onChange={e => setNote(e.target.value)}
      >
      </textarea>
      <button className="btn-push">Save Event</button>
    </form>
  )
}