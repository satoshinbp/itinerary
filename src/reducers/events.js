const eventsReducerDefaultState = []

export default (state = eventsReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EVENT':
      return [
        ...state,
        action.event
      ]
    case 'EDIT_EVENT':
      return state.map((event) => {
        if (event.id === action.id) {
          return {
            ...event,
            ...action.updates
          }
        } else {
          return event
        }
      })
    case 'REMOVE_EVENT':
      return state.filter(({ id }) => id !== action.id)
    default:
      return state
  }
}