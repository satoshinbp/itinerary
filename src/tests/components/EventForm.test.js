import React from 'react'
import { shallow } from 'enzyme'
import moment from 'moment'
import EventForm from '../../components/EventForm'
import events from '../fixtures/events'

Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
})

let setState, useStateSpy

beforeEach(() => {
  setState = jest.fn()
  useStateSpy = jest.spyOn(React, 'useState')
  useStateSpy.mockImplementation(init => [init, setState])
})

test('should render EventForm correctly', () => {
  const wrapper = shallow(<EventForm />)
  expect(wrapper).toMatchSnapshot()
})

test('should render EventForm correctly with trip data', () => {
  const wrapper = shallow(<EventForm event={events[0]} />)
  expect(wrapper).toMatchSnapshot()
})

test('should render error for invalid form submission', () => {
  const wrapper = shallow(<EventForm />)
  wrapper.find('form').simulate('submit', {
    preventDefault: () => {}
  })
  expect(setState).toBeCalledWith(expect.any(String))
})

test('should call onSubmit prop for valid form submission', () => {
  const onSubmitSpy = jest.fn()
  const wrapper = shallow(<EventForm onSubmit={onSubmitSpy} event={events[0]} />)
  wrapper.find('form').simulate('submit', {
    preventDefault: () => {}
  })
  expect(setState).toBeCalledWith('')
  expect(onSubmitSpy).lastCalledWith({
    title: events[0].title,
    date: events[0].date,
    startTime: events[0].startTime,
    endTime: events[0].endTime,
    location: events[0].location,
    note: events[0].note
  })
})

test('should set title on input change', () => {
  const value = 'New Title'
  const wrapper = shallow(<EventForm />)
  wrapper.find('input').at(0).simulate('change', {
    target: { value }
  })
  expect(setState).toBeCalledWith(value)
})

test('should set location on input change', () => {
  const value = 'New Location'
  const wrapper = shallow(<EventForm />)
  wrapper.find('input').at(1).simulate('change', {
    target: { value }
  })
  expect(setState).toBeCalledWith(value)
})

test('should set note on textarea change', () => {
  const value = 'New Note'
  const wrapper = shallow(<EventForm />)
  wrapper.find('textarea').simulate('change', {
    target: { value }
  })
  expect(setState).toBeCalledWith(value)
})

test('should set date on date change', () => {
  const date = moment(0)
  const wrapper = shallow(<EventForm />)
  wrapper.find('withStyles(SingleDatePicker)').prop('onDateChange')(date)
  expect(setState).toBeCalledWith(moment(date).startOf('day'))
})

test('should set calendar focus on change', () => {
  const focused = true
  const wrapper = shallow(<EventForm />)
  wrapper.find('withStyles(SingleDatePicker)').prop('onFocusChange')({ focused })
  expect(setState).toBeCalledWith(focused)
})

test('should set start time on change', () => {
  const time = moment().hour(23).minutes(59)
  const wrapper = shallow(<EventForm event={events[0]} />)
  wrapper.find('PickerWithState').at(0).prop('onChange')(time)
  expect(setState).nthCalledWith(1, time)
  expect(setState).nthCalledWith(2, time)
})

test('should set end time on change', () => {
  const time = moment().hour(1).minutes(0)
  const wrapper = shallow(<EventForm />)
  wrapper.find('PickerWithState').at(1).prop('onChange')(time)
  expect(setState).nthCalledWith(1, time)
  expect(setState).nthCalledWith(2, time)
})