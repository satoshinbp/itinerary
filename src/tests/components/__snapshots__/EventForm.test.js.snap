// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EventForm correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="title"
    type="text"
    value=""
  />
  <div
    className="date-time-picker"
  >
    <withStyles(SingleDatePicker)
      anchorDirection="left"
      appendToBody={false}
      block={false}
      calendarInfoPosition="bottom"
      customCloseIcon={null}
      customInputIcon={null}
      date={"1970-01-01T00:00:00.000Z"}
      dayPickerNavigationInlineStyles={null}
      daySize={39}
      disableScroll={false}
      disabled={false}
      displayFormat="YYYY/MM/DD"
      enableOutsideDays={false}
      firstDayOfWeek={null}
      focused={null}
      hideKeyboardShortcutsPanel={false}
      horizontalMargin={0}
      horizontalMonthPadding={13}
      id="date"
      initialVisibleMonth={null}
      inputIconPosition="before"
      isDayBlocked={[Function]}
      isDayHighlighted={[Function]}
      isOutsideRange={[Function]}
      isRTL={false}
      keepFocusOnInput={false}
      keepOpenOnDateSelect={false}
      monthFormat="MMMM YYYY"
      navNext={null}
      navPosition="navPositionTop"
      navPrev={null}
      noBorder={false}
      numberOfMonths={1}
      onClose={[Function]}
      onDateChange={[Function]}
      onFocusChange={[Function]}
      onNextMonthClick={[Function]}
      onPrevMonthClick={[Function]}
      openDirection="down"
      orientation="horizontal"
      phrases={
        Object {
          "calendarLabel": "Calendar",
          "chooseAvailableDate": [Function],
          "clearDate": "Clear Date",
          "closeDatePicker": "Close",
          "dateIsSelected": [Function],
          "dateIsUnavailable": [Function],
          "enterKey": "Enter key",
          "escape": "Escape key",
          "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
          "homeEnd": "Home and end keys",
          "jumpToNextMonth": "Move forward to switch to the next month.",
          "jumpToPrevMonth": "Move backward to switch to the previous month.",
          "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardShortcuts": "Keyboard Shortcuts",
          "leftArrowRightArrow": "Right and left arrow keys",
          "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
          "moveFocusByOneMonth": "Switch months.",
          "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
          "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
          "openThisPanel": "Open this panel.",
          "pageUpPageDown": "page up and page down keys",
          "questionMark": "Question mark",
          "returnFocusToInput": "Return to the date input field.",
          "roleDescription": "datepicker",
          "selectFocusedDate": "Select the date in focus.",
          "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
          "upArrowDownArrow": "up and down arrow keys",
        }
      }
      placeholder="Date"
      readOnly={false}
      regular={false}
      renderCalendarInfo={null}
      renderDayContents={null}
      renderMonthElement={null}
      renderMonthText={null}
      renderNavNextButton={null}
      renderNavPrevButton={null}
      renderWeekHeaderElement={null}
      reopenPickerOnClearDate={false}
      required={false}
      screenReaderInputMessage=""
      showClearDate={false}
      showDefaultInputIcon={false}
      small={false}
      verticalHeight={null}
      verticalSpacing={22}
      weekDayFormat="dd"
      withFullScreenPortal={false}
      withPortal={false}
    />
  </div>
  <div
    className="date-time-picker"
  >
    <MuiPickersUtilsProvider
      utils={[Function]}
    >
      <PickerWithState
        ampm={true}
        invalidDateMessage="Invalid Time Format"
        label="Start Time"
        onChange={[Function]}
        openTo="hours"
        value={"1970-01-01T03:00:00.000Z"}
        views={
          Array [
            "hours",
            "minutes",
          ]
        }
      />
    </MuiPickersUtilsProvider>
    <MuiPickersUtilsProvider
      utils={[Function]}
    >
      <PickerWithState
        ampm={true}
        invalidDateMessage="Invalid Time Format"
        label="End Time"
        onChange={[Function]}
        openTo="hours"
        value={"1970-01-01T03:00:00.000Z"}
        views={
          Array [
            "hours",
            "minutes",
          ]
        }
      />
    </MuiPickersUtilsProvider>
  </div>
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="location"
    type="text"
    value=""
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add a note for your event (optional)"
    value=""
  />
  <button
    className="btn-push"
  >
    Save Event
  </button>
</form>
`;

exports[`should render EventForm correctly with trip data 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="title"
    type="text"
    value="Shopping"
  />
  <div
    className="date-time-picker"
  >
    <withStyles(SingleDatePicker)
      anchorDirection="left"
      appendToBody={false}
      block={false}
      calendarInfoPosition="bottom"
      customCloseIcon={null}
      customInputIcon={null}
      date={"1970-01-02T00:00:00.000Z"}
      dayPickerNavigationInlineStyles={null}
      daySize={39}
      disableScroll={false}
      disabled={false}
      displayFormat="YYYY/MM/DD"
      enableOutsideDays={false}
      firstDayOfWeek={null}
      focused={null}
      hideKeyboardShortcutsPanel={false}
      horizontalMargin={0}
      horizontalMonthPadding={13}
      id="date"
      initialVisibleMonth={null}
      inputIconPosition="before"
      isDayBlocked={[Function]}
      isDayHighlighted={[Function]}
      isOutsideRange={[Function]}
      isRTL={false}
      keepFocusOnInput={false}
      keepOpenOnDateSelect={false}
      monthFormat="MMMM YYYY"
      navNext={null}
      navPosition="navPositionTop"
      navPrev={null}
      noBorder={false}
      numberOfMonths={1}
      onClose={[Function]}
      onDateChange={[Function]}
      onFocusChange={[Function]}
      onNextMonthClick={[Function]}
      onPrevMonthClick={[Function]}
      openDirection="down"
      orientation="horizontal"
      phrases={
        Object {
          "calendarLabel": "Calendar",
          "chooseAvailableDate": [Function],
          "clearDate": "Clear Date",
          "closeDatePicker": "Close",
          "dateIsSelected": [Function],
          "dateIsUnavailable": [Function],
          "enterKey": "Enter key",
          "escape": "Escape key",
          "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
          "homeEnd": "Home and end keys",
          "jumpToNextMonth": "Move forward to switch to the next month.",
          "jumpToPrevMonth": "Move backward to switch to the previous month.",
          "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardShortcuts": "Keyboard Shortcuts",
          "leftArrowRightArrow": "Right and left arrow keys",
          "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
          "moveFocusByOneMonth": "Switch months.",
          "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
          "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
          "openThisPanel": "Open this panel.",
          "pageUpPageDown": "page up and page down keys",
          "questionMark": "Question mark",
          "returnFocusToInput": "Return to the date input field.",
          "roleDescription": "datepicker",
          "selectFocusedDate": "Select the date in focus.",
          "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
          "upArrowDownArrow": "up and down arrow keys",
        }
      }
      placeholder="Date"
      readOnly={false}
      regular={false}
      renderCalendarInfo={null}
      renderDayContents={null}
      renderMonthElement={null}
      renderMonthText={null}
      renderNavNextButton={null}
      renderNavPrevButton={null}
      renderWeekHeaderElement={null}
      reopenPickerOnClearDate={false}
      required={false}
      screenReaderInputMessage=""
      showClearDate={false}
      showDefaultInputIcon={false}
      small={false}
      verticalHeight={null}
      verticalSpacing={22}
      weekDayFormat="dd"
      withFullScreenPortal={false}
      withPortal={false}
    />
  </div>
  <div
    className="date-time-picker"
  >
    <MuiPickersUtilsProvider
      utils={[Function]}
    >
      <PickerWithState
        ampm={true}
        invalidDateMessage="Invalid Time Format"
        label="Start Time"
        onChange={[Function]}
        openTo="hours"
        value={"1970-01-01T04:00:00.000Z"}
        views={
          Array [
            "hours",
            "minutes",
          ]
        }
      />
    </MuiPickersUtilsProvider>
    <MuiPickersUtilsProvider
      utils={[Function]}
    >
      <PickerWithState
        ampm={true}
        invalidDateMessage="Invalid Time Format"
        label="End Time"
        onChange={[Function]}
        openTo="hours"
        value={"1970-01-01T07:00:00.000Z"}
        views={
          Array [
            "hours",
            "minutes",
          ]
        }
      />
    </MuiPickersUtilsProvider>
  </div>
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="location"
    type="text"
    value="Waikiki"
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add a note for your event (optional)"
    value=""
  />
  <button
    className="btn-push"
  >
    Save Event
  </button>
</form>
`;
